src/riot-rs-runqueue/src/runqueue.rs

```diff
@@ -98,34 +98,136 @@ impl<const N_QUEUES: usize, const N_THREADS: usize> RunQueue<{ N_QUEUES }, { N_T
+    pub fn peek_next(&self) -> Option<(ThreadId, RunqueueId)> {
+        let rq_ffs = ffs(self.bitcache);
+        if rq_ffs == 0 {
+            return None;
+        }
+        let rq = rq_ffs as u8 - 1;
+        match self.queues.peek_head(rq) {
+            Some(id) => Some((ThreadId::new(id), RunqueueId::new(rq))),
+            None => None,
+        }
+    }
+
+    /// Pop the thread that should run next.
+    ///
+    /// Pops the next runnable thread of
+    /// the runqueue with the highest index.
+    pub fn pop_next(&mut self) -> Option<ThreadId> {
+        let rq_ffs = ffs(self.bitcache);
+        if rq_ffs == 0 {
+            return None;
+        }
+        let rq = (rq_ffs - 1) as u8;
+        let head = match self.queues.pop_head(rq) {
+            Some(head) => Some(ThreadId(head)),
+            None => return None,
+        };
+        if self.queues.is_empty(rq) {
+            self.bitcache &= !(1 << rq);
+        }
+        head
+    }
+
     /// Advances runqueue number `rq`.
     ///
     /// This is used to "yield" to another thread of *the same* priority.
-    pub fn advance(&mut self, rq: RunqueueId) {
+    ///
+    /// Returns `false` if the operation had no effect, i.e. when the runqueue
+    /// is empty or only contains a single thread.
+    pub fn advance(&mut self, rq: RunqueueId) -> bool {
         debug_assert!((usize::from(rq)) < N_QUEUES);
         self.queues.advance(rq.0)
     }
+
+    /// Removes thread with pid `n`.
+    pub fn del(&mut self, n: ThreadId) {
+        if let Some(empty_runqueue) = self.queues.del(n.0) {
+            self.bitcache &= !(1 << empty_runqueue);
+        }
+    }
+
+    /// Checks if a runqueue is empty.
+    pub fn is_empty(&mut self, rq: RunqueueId) -> bool {
+        debug_assert!((rq.0 as usize) < N_QUEUES);
+        self.queues.is_empty(rq.0)
+    }
+
+    /// Iterate over all runqueues, starting after thread `start` in runqueue `rq`.
+    ///
+    /// The `start` is not included in the iterator.
+    pub fn iter_from(&self, start: ThreadId, rq: RunqueueId) -> RunQueueIter<N_QUEUES, N_THREADS> {
+        RunQueueIter {
+            prev: start.0,
+            rq_head: self.queues.peek_head(rq.0),
+            bitcache: self.bitcache % (1 << (rq.0 + 1)),
+            queues: &self.queues,
+        }
+    }
+}
+
+fn ffs(val: usize) -> u32 {
+    USIZE_BITS as u32 - val.leading_zeros()
+}
+
+/// Iterator over all queues in a [`RunQueue`].
+///
+/// It starts from the highest priority queue and switches to the next lower
+/// priority runqueue after circling through the queue once.
+pub struct RunQueueIter<'a, const N_QUEUES: usize, const N_THREADS: usize> {
+    queues: &'a clist::CList<N_QUEUES, N_THREADS>,
+    // Predecessor in the circular runqueue list.
+    prev: u8,
+    // Head of the currently iterated runqueue.
+    rq_head: Option<u8>,
+    // Bitcache with the remaining runqueues that have to be iterated.
+    bitcache: usize,
+}
```

Comment:
I slightly dislike that I had to add this rather complex iterator here.
It's needed for `core-affinity` because we need to "get the next thread from the runqueue that fulfills condition x". The best way to do that that I came up with is with `Iterator` + `filter`. But I am happy to hear alternative ideas.

Reply:
And unfortunately it also fails the hax check because`Iterator` isn't supported. I guess i could work around it to just manually implement a filter function for the runqueue...

Reply:
Moved to #460.